File: backend/app/schemas/timeline.py
from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel


class TimelineAssetResponse(BaseModel):
    id: str
    project_id: str
    asset_type: str
    filename: str
    file_url: str
    file_size: int
    duration_ms: Optional[int] = None

    class Config:
        from_attributes = True


class TimelineClipData(BaseModel):
    generated_video_id: Optional[str] = None
    asset_id: Optional[str] = None
    track_type: str = "video"
    track_index: int = 0
    position_ms: int
    duration_ms: int
    sort_order: int
    label: Optional[str] = None


class TimelineClipResponse(BaseModel):
    id: str
    generated_video_id: Optional[str] = None
    asset_id: Optional[str] = None
    track_type: str
    track_index: int
    position_ms: int
    duration_ms: int
    sort_order: int
    label: Optional[str] = None
    # Resolved URLs
    video_url: Optional[str] = None
    thumbnail_url: Optional[str] = None
    filename: Optional[str] = None

    class Config:
        from_attributes = True


class TimelineSaveRequest(BaseModel):
    clips: List[TimelineClipData]


class TimelineResponse(BaseModel):
    project_id: str
    clips: List[TimelineClipResponse]
    assets: List[TimelineAssetResponse] = []

File: frontend/src/types/timeline.ts
export interface TalkingHeadTask {
  id: string
  project_id: string
  shot_index: number | null

  model_image_id: string
  model_image_url: string | null
  bg_material_id: string | null
  bg_thumbnail_url: string | null

  composite_status: 'pending' | 'processing' | 'completed' | 'failed'
  composite_preview_url: string | null

  motion_prompt: string | null
  audio_segment_url: string | null
  audio_start_ms: number | null
  audio_end_ms: number | null

  lipsync_status: 'pending' | 'processing' | 'completed' | 'failed'
  video_url: string | null
  thumbnail_url: string | null
  duration_seconds: number | null
  error_message: string | null

  created_at: string
  completed_at: string | null
}

export interface TimelineAsset {
  id: string
  project_id: string
  asset_type: 'video' | 'audio' | 'subtitle'
  filename: string
  file_url: string
  file_size: number
  duration_ms: number | null
}

export interface TimelineClip {
  id: string
  generated_video_id: string | null
  asset_id: string | null
  track_type: 'video' | 'audio' | 'subtitle'
  track_index: number
  position_ms: number
  duration_ms: number
  sort_order: number
  label: string | null
  video_url: string | null
  thumbnail_url: string | null
  filename: string | null
}

export interface Timeline {
  project_id: string
  clips: TimelineClip[]
  assets: TimelineAsset[]
}
